// robot/index.js
var sleep = require('sleep');
var io = require('socket.io-client');
var socket;

module.exports = {
	connect: function(s, k, a) {
		socket = io.connect('http://localhost:3000');
		socket.on('connect',function() {
			console.log('Client has connected to the server!');
			var trampoline = k(s);
			while (trampoline){
				trampoline = trampoline();
			}
		});
	},
	sleep: sleep.sleep,
	delay: function(action, seconds) {
		var waitTill = new Date(
			new Date().getTime() + 2//(seconds * 1000)
		);
		while(waitTill > new Date()){};
		return action();
	},
	moveForward: function(s, k, a) {
	    socket.emit('start');
	    socket.on('done', function() {
	    	socket.removeListener('done');
	    	sleep.sleep(1);
			console.log('robot sent me back "done"');
			var trampoline = k(s);
			while (trampoline){
				trampoline = trampoline();
			}
		})
	},
	turnRight: function() {
	    socket.emit('right');
	},
	turnLeft: function() {
	    socket.emit('left');
	},
	moveReverse: function() {
	    socket.emit('reverse');
	},
	stop: function() {
	    socket.emit('stop');
	},
	reverseRight: function() {
	    socket.emit('reverse-right');
	}
};
