// robot/index.js
var sleep = require('sleep');
var io = require('socket.io-client');
var socket;

function done(s, k, a, seconds) {
	var seconds = seconds ? seconds : 1;
    socket.on('done', function() {
    	socket.removeListener('done');
    	sleep.usleep(seconds*1000000);
		// console.log('robot sent me back "done"');
		var trampoline = k(s);
		while (trampoline){
			trampoline = trampoline();
		}
	})
}

module.exports = {
	connect: function(s, k, a) {
		socket = io.connect('http://localhost:3000');
		socket.on('connect',function() {
			console.log('Client has connected to the server!');
			var trampoline = k(s);
			while (trampoline){
				trampoline = trampoline();
			}
		});
	},
	sleep: sleep.sleep,
	range: function(a, b) {
		if (a<b) {
			var lst = [];
			for (var i=a; i<b; i++) {
				lst.push(i);
			}
			return lst;
		} else {
			var lst = [];
			for (var i=a; i>b; i--) {
				lst.push(i);
			}
			return lst;
		}
	},
	delay: function(action, seconds) {
		var waitTill = new Date(
			new Date().getTime() + 2//(seconds * 1000)
		);
		while(waitTill > new Date()){};
		return action();
	},
	moveForward: function(s, k, a, seconds) {
	    socket.emit('start');
		done(s, k, a, seconds);
	},
	turnRight: function(s, k, a, seconds) {
	    socket.emit('right');
		done(s, k, a, seconds);
	},
	turnLeft: function(s, k, a, seconds) {
	    socket.emit('left');
		done(s, k, a, seconds);
	},
	moveReverse: function(s, k, a, seconds) {
	    socket.emit('reverse');
		done(s, k, a, seconds);
	},
	stop: function(s, k, a, seconds) {
	    socket.emit('stop');
		done(s, k, a, seconds);
	},
	reverseRight: function(s, k, a, seconds) {
	    socket.emit('reverse-right');
		done(s, k, a, seconds);
	},
	sendMap: function(s, k, a, aMap) {
		socket.emit('send-map', aMap);
	    socket.on('done', function() {
	    	socket.removeListener('done');
			var trampoline = k(s);
			while (trampoline){
				trampoline = trampoline();
			}
		})
	},
	sendPosition: function(s, k, a, position) {
		socket.emit('send-position', position);
	    socket.on('done', function() {
	    	socket.removeListener('done');
			var trampoline = k(s);
			while (trampoline){
				trampoline = trampoline();
			}
		})
	},
	getLeftDistance: function(s, k, a) {
		socket.emit('get-left-distance');
	    socket.on('done', function(leftDistance) {
	    	socket.removeListener('done');
	    	// sleep.usleep(seconds*1000000);
			// console.log('robot sent me back "left distance"');
			var trampoline = k(s, leftDistance);
			while (trampoline){
				trampoline = trampoline();
			}
		})
	},
	getRightDistance: function(s, k, a) {
		socket.emit('get-right-distance');
	    socket.on('done', function(rightDistance) {
	    	socket.removeListener('done');
	    	// sleep.usleep(seconds*1000000);
			// console.log('robot sent me back "right distance"');
			var trampoline = k(s, rightDistance);
			while (trampoline){
				trampoline = trampoline();
			}
		})
	}
};
